### PR TITLE
[NEW FILTER] FYara

### PR DESCRIPTION
**Pull request by frikilax**
# :sparkles: Implement new filter fyara
:bangbang: Once all the **checklist** is **done** you have to:
  * **stash merge** this pull request
  * **delete** the corresponding **branch**
  * **close** the associated **issue**

## :page_with_curl: Type of change

Please delete options that are not relevant.

**New feature**: non-breaking change which adds functionality.

## :bulb: Related Issue(s)

- Replaces #84 

### Filters
- **[NEW FEATURE]** add yara filter
  - possibility to scan arbitrary data
  - possibility to read data as-is, or decode hex or base64-encoded data
  - possibility to give a list of files to include during yara rules compilation
  - possibility to run yara engine in "fastmode" and to define a "timeout" (see documentation)

### Toolkit
- add Yara engine utilities
- add encoders utilities to encode/decode hex/base64

### Tests
- add functional tests for fyara

### CMake
- update CmakeLists to compile fyara by default

### Docker
- update Armadillo to version 9.900.x
- add yara as a compilation step (current ubuntu 18.04 version is too old for features)

### HBSD 12.1
- Redis (5.0.9)
- Boost (1.72.0_2)
- g++ (9.3.0)
- CMake (3.17.3)
- Python (3.7)
- yara (3.11.0)

### Ubuntu (18.04)
- Redis (4.0.9)
- Boost (1.71.0)
- g++ (7.5.0)
- CMake (3.10.2)
- Python (3.6)
- Valgrind (3.13.0)
- yara (3.11.0)

## :heavy_check_mark: Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] (**If new filter**) I have added corresponding page to the documentation
- [x] (**If other changes**) I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

</br>

- [x] :raising_hand: **I certify on my honor that all the information provided is true, and I've done all I can to deliver a high quality code**

### CODE REVIEW

File: toolkit/Yara.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        void YaraEngine::AddRuleToMatch(YR_RULE *rule) {
+            _rule_match_list.insert(rule);
+        }
+    } // namespace toolkit
+} // namespace darwin
+
+
+namespace darwin {
+
+    namespace toolkit {
```
Comment (HugoSoszynski) [inline toolkit/Yara.cpp:146]: No need to close and reopen the namespaces.

```suggestion
```

File: toolkit/Yara.hpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            YaraEngine(bool fastmode = true, int timeout = 0);
+            ~YaraEngine();
+
+            bool Init(YR_RULES *rules);
+            int ScanData(std::vector<unsigned char> &data, unsigned int &certitude);
+            rapidjson::Document GetResults();
+
+        private:
+            static int ScanOrImportCallback(int message, void *messageData, void *userData);
+            void AddRuleToMatch(YR_RULE *rule);
```
Comment (HugoSoszynski) [inline toolkit/Yara.hpp:28]: It would be nice to include the Doxygen formated documentation.

File: toolkit/Yara.hpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                ERROR}; //status of compiler
+
+            YaraCompiler() = default;
+            ~YaraCompiler();
+
+            bool Init();
+            bool AddRuleFile(FILE *file, std::string nameSpace, std::string filename);
+            bool CompileRules();
+            Status GetStatus();
+            std::shared_ptr<YaraEngine> GetEngine(bool fastmode = false, int timeout = 0);
```
Comment (HugoSoszynski) [inline toolkit/Yara.hpp:52]: It would be nice to include the Doxygen formated documentation

File: toolkit/Yara.hpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        private:
+            static void YaraErrorCallback(int errorLevel, const char *filename, int lineNumber, const char *message, void *userData);
+
+        private:
+            YR_COMPILER *_compiler = nullptr; //Yara compiler
+            YR_RULES *_rules = nullptr; //Yara compiled rules
```
Comment (HugoSoszynski) [inline toolkit/Yara.hpp:59]: Why not using some unique_ptr and use the `std::unique_ptr::get()` method when interfacing with C ?

File: CMakeLists.txt [hunk 5]
```diff
@@ -101,6 +136,8 @@ set(
         toolkit/Files.cpp toolkit/Files.hpp
         toolkit/RedisManager.cpp toolkit/RedisManager.hpp
         toolkit/FileManager.cpp toolkit/FileManager.hpp
+        toolkit/xxhash.h toolkit/xxhash.hpp
+        toolkit/lru_cache.hpp
```
[the remark this answers is not in the document]
Comment (frikilax) [reply CMakeLists.txt]: Done in 2b853cb

File: toolkit/Yara.cpp (hunk 1 above)
Comment (frikilax) [reply toolkit/Yara.cpp]: Done in 3e21146

File: toolkit/Yara.hpp (hunk 2 above)
Comment (frikilax) [reply toolkit/Yara.hpp]: Done in ef19641

File: toolkit/Yara.hpp (hunk 3 above)
Comment (frikilax) [reply toolkit/Yara.hpp]: Done in ef19641

File: toolkit/Yara.hpp (hunk 4 above)
Comment (frikilax) [reply toolkit/Yara.hpp:110]: I admit being more familiar with C-style pointers, and as data pointed is not trivially freed the interest of a C++ style pointer decreases

**Pull request merged**